lib/global-styles-and-settings.php

```diff
@@ -261,8 +266,16 @@ function gutenberg_get_global_styles( $path = array(), $context = array() ) {
 	if ( isset( $context['origin'] ) && 'base' === $context['origin'] ) {
 		$origin = 'theme';
 	}
-	$styles = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data( $origin )->get_raw_data()['styles'];
 
+	$resolve_variables = isset( $context['transforms'] )
+	&& is_array( $context['transforms'] )
+	&& in_array( 'resolve-variables', $context['transforms'], true );
+
+	if ( $resolve_variables ) {
```

Comment:
This works, but the idea is to have more transforms available. How would they be composable when we have two of them? Note https://github.com/WordPress/gutenberg/pull/50484#discussion_r1192502874 that we can have 4 use cases:

1. no transforms
2. only `resolve-variables`
3. only `resolve-references`
4. both `resolve-variables` and `resolve-references`

I worry that if we create a method in the `WP_Theme_JSON_Gutenberg` class for each of the use cases from 2-4 implementing the 4th becomes specially complex to read.

What if we make the transformations composable and standalone operations? For example, we could have one function for transformation: 

```php
/**
 * @access private 
 * @param array $styles The style nodes to transform.
 * @return The transformed style nodes.
 */
function _gutenberg_transform_theme_json_styles_resolve_variables( $styles ) {}
```

This way, the code at `wp_get_global_styles` becomes much simpler as we evolve it:

```php
$styles = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data( $origin )->get_raw_data()['styles'];

if ( $resolve_variables ) {
  $styles = _gutenberg_transform_theme_json_styles_resolve_variables( $styles );
}

if ( $resolve_references ) {
  $styles = _gutenberg_transform_theme_json_styles_resolve_references( $styles );
}

return $styles;
```

What do you think?

Reply:
If I understand you correctly, you are suggesting to move the logic for these transformation from private class methods (i.e. `WP_Theme_JSON_Gutenberg::get_styles_with_values`) out of the `WP_Theme_JSON_Gutenberg` and into a function like `_gutenberg_transform_theme_json_styles_resolve_variables` but annotate it as private? And in both cases we should resolve the value for `$resolve_variables` and later `$resolve_references` inside `wp_get_global_styles`, right?

It'll prevent the class from growing bigger for sure, but I'm not getting how it will make it more readable.
The only advantage I see in your suggestion is having **composable** functions like `_gutenberg_transform_theme_json_styles_resolve_variables`, while `WP_Theme_JSON_Gutenberg::get_styles_with_values` is by nature static, and we can make it static and allow passing the `$styles` from outside, like the following:

```
public static function get_styles_with_values( $styles = null ) {
     ....
     $_styles = $styles && is_array($styles) ? $styles : self::get_raw_data()['styles'];
      ....
     return self::convert_variables_to_value( $styles, $vars );
}
```

That said I'm fine either way.

Reply:
Yeah, either a function or a public static method of the private class `WP_Theme_JSON_Gutenberg` is fine by me. Perhaps I'm leaning toward separate functions because of the size of the class, though it's fine to start with the static method. Not opinionated about that.

What I care about is that the the transformation is a pure operation that takes `$styles` as input and returns the transformed `$styles` as output: `$styles = transformation( $styles )`.

Reply:
Looking at the existing code, I have a suggestion to make, what if we do this?

```php
// ...
$merged_data = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data( $origin );
if ( $resolve_variables ) {
    $merged_data = WP_Theme_JSON_Gutenberg::transform_resolve_variables( $merged_data );
}
$styles = $merged_data->get_raw_data()['styles'];
// ...
```

Notes:

- The code that executes the transformations could take a `$theme_json` as input and return the same structure. This is so consumers don't need to know the details of the inner operation (does it use settings? etc.?), simplifying the consumer code. It'd also make all transformations use the same API. Note this is what we already do for the existing `resolve_custom_css_format`, so building on that would be great.
- Ideally, the name of the method in the theme.json class is representative of the name of the parameter for the transformation. For example, if the parameter is called `resolve-variables`, I'd expect the method to be called something like `resolve_variables`.

Reply:
addressed in fff3c43707e3e8f36ed7b9c15427b34a4a93f7ac